clib/sList.mli

```diff
@@ -0,0 +1,73 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(** Sparse lists. *)
+
+(** {5 Constructors} *)
+
+type +'a t = private
+| Nil
+| Cons of 'a * 'a t
+| Default of int * 'a t
+(** ['a t] is an efficient representation of ['a option list]. *)
+
+val empty : 'a t
+(** The empty list. *)
+
+val cons : 'a -> 'a t -> 'a t
+(** Isomorphic to [Some x :: l]. *)
+
+val default : 'a t -> 'a t
+(** Isomorphic to [None :: l]. *)
+
+val cons_opt : 'a option -> 'a t -> 'a t
+(** {!cons} if [Some], {!default} otherwise *)
+
+val defaultn : int -> 'a t -> 'a t
+(** Iterated variant of [default]. *)
+
+(** {5 Destructor} *)
+
+val view : 'a t -> ('a option * 'a t) option
+
+val is_empty : 'a t -> bool
+
+(** {5 Usual list-like operators} *)
+
+val length : 'a t -> int
+
+val equal : ('a option -> 'b option -> bool) -> 'a t -> 'b t -> bool
+
+val compare : ('a option -> 'a option -> int) -> 'a t -> 'a t -> int
```

Comment:
Is it ever useful to pass these something that's not `Option.equal something` / `Option.compare something`?

Reply:
Nobody really uses those functions anyways.

Reply:
I would still prefer to have the API be `('a -> 'b -> bool) -> 'a t -> 'b t -> bool` (resp `int` for compare)